# Stage I - Data Understanding and Linking

## Member Task- Raveena Arasikere Rakesh

## Task-2 M2.1

The solution to this task is at location- '././../../../doc/Opioid-stage 1.docx'



### Task-2 M2.2

#### Reading the dataset

In [37]:
import pandas as pd
file1=pd.read_csv('../data/Underlying Cause of Death-County-2019.txt')
file1.head(5)#displays only top-5 elemets

Notes\t"County"\t"County Code"\t"Drug/Alcohol Induced Cause"\t"Drug/Alcohol Induced Cause Code"\tDeaths\tPopulation\tCrude Rate
\t"Autauga County   AL"\t"01001"\t"Drug poisonings (overdose) Uni...                                                                             
\t"Autauga County   AL"\t"01001"\t"Drug poisonings (overdose) Sui...                                                                             
\t"Baldwin County   AL"\t"01003"\t"Drug poisonings (overdose) Uni...                                                                             
\t"Baldwin County   AL"\t"01003"\t"Drug poisonings (overdose) Sui...                                                                             
\t"Baldwin County   AL"\t"01003"\t"Drug poisonings (overdose) Und...

<li> From the above rendered output we can see that there are a lot of '\t' so we need to either use a delimeter or a seperater sep function to display the dataframe in an organised format.

In [38]:
file1=pd.read_csv('../data/Underlying Cause of Death-County-2019.txt',sep='\t')
file1.head()

Notes              County  County Code  \
0    NaN  Autauga County, AL         1001   
1    NaN  Autauga County, AL         1001   
2    NaN  Baldwin County, AL         1003   
3    NaN  Baldwin County, AL         1003   
4    NaN  Baldwin County, AL         1003   

                          Drug/Alcohol Induced Cause  \
0  Drug poisonings (overdose) Unintentional (X40-...   
1       Drug poisonings (overdose) Suicide (X60-X64)   
2  Drug poisonings (overdose) Unintentional (X40-...   
3       Drug poisonings (overdose) Suicide (X60-X64)   
4  Drug poisonings (overdose) Undetermined (Y10-Y14)   

  Drug/Alcohol Induced Cause Code  Deaths  Population  Crude Rate  
0                              D1      69     1087149         6.3  
1                              D2      14     1087149  Unreliable  
2                              D1     424     3758097        11.3  
3                              D2      71     3758097         1.9  
4                              D4      19     3758097  Unreliable

<li> Now that the dataframe is rendered in a readable format we can now perform some data cleaning operations like replacing the NAN values,negative values if there exists any

In [39]:
file1.isna().head()

Notes  County  County Code  Drug/Alcohol Induced Cause  \
0   True   False        False                       False   
1   True   False        False                       False   
2   True   False        False                       False   
3   True   False        False                       False   
4   True   False        False                       False   

   Drug/Alcohol Induced Cause Code  Deaths  Population  Crude Rate  
0                            False   False       False       False  
1                            False   False       False       False  
2                            False   False       False       False  
3                            False   False       False       False  
4                            False   False       False       False

<li>Since, there exists some NAN values we are going to replace them with 0

In [40]:
file1.fillna(0)

Notes                 County  County Code  \
0       0.0     Autauga County, AL         1001   
1       0.0     Autauga County, AL         1001   
2       0.0     Baldwin County, AL         1003   
3       0.0     Baldwin County, AL         1003   
4       0.0     Baldwin County, AL         1003   
...     ...                    ...          ...   
5536    0.0  Sweetwater County, WY        56037   
5537    0.0       Teton County, WY        56039   
5538    0.0       Uinta County, WY        56041   
5539    0.0       Uinta County, WY        56041   
5540    0.0    Washakie County, WY        56043   

                             Drug/Alcohol Induced Cause  \
0     Drug poisonings (overdose) Unintentional (X40-...   
1          Drug poisonings (overdose) Suicide (X60-X64)   
2     Drug poisonings (overdose) Unintentional (X40-...   
3          Drug poisonings (overdose) Suicide (X60-X64)   
4     Drug poisonings (overdose) Undetermined (Y10-Y14)   
...                                                 ...   
5536       Drug poisonings (overdose) Suicide (X60-X64)   
5537  Drug poisonings (overdose) Unintentional (X40-...   
5538  Drug poisonings (overdose) Unintentional (X40-...   
5539  Drug poisonings (overdose) Undetermined (Y10-Y14)   
5540  Drug poisonings (overdose) Unintentional (X40-...   

     Drug/Alcohol Induced Cause Code  Deaths  Population  Crude Rate  
0                                 D1      69     1087149         6.3  
1                                 D2      14     1087149  Unreliable  
2                                 D1     424     3758097        11.3  
3                                 D2      71     3758097         1.9  
4                                 D4      19     3758097  Unreliable  
...                              ...     ...         ...         ...  
5536                              D2      15      873221  Unreliable  
5537                              D1      13      440125  Unreliable  
5538                              D1      51      426347        12.0  
5539                              D4      10      426347  Unreliable  
5540                              D1      14      172114  Unreliable  

[5541 rows x 8 columns]

In [41]:
(file1.Population<0).any()#checking for negative values in the poulation field,because negative population is of no use

False

###### Reading file-2

In [42]:
file2=pd.read_csv('../data/County_Health_Ranking.csv')
file2.head()

State FIPS Code  County FIPS Code  5-digit FIPS Code State Abbreviation  \
0                1                 0               1000                 AL   
1                1                 1               1001                 AL   
2                1                 3               1003                 AL   
3                1                 5               1005                 AL   
4                1                 7               1007                 AL   

             Name  Release Year  County Ranked (Yes=1/No=0)  \
0         Alabama          2019                         NaN   
1  Autauga County          2019                         1.0   
2  Baldwin County          2019                         1.0   
3  Barbour County          2019                         1.0   
4     Bibb County          2019                         1.0   

   Premature death raw value  Premature death numerator  \
0                9917.232898                    80440.0   
1                8824.057123                      815.0   
2                7224.632160                     2827.0   
3                9586.165037                      451.0   
4               11783.543680                      445.0   

   Premature death denominator  ...  Male population 18-44 raw value  \
0                   13636816.0  ...                              NaN   
1                     156132.0  ...                              NaN   
2                     576496.0  ...                              NaN   
3                      72222.0  ...                              NaN   
4                      63653.0  ...                              NaN   

   Male population 45-64 raw value  Male population 65+ raw value  \
0                              NaN                            NaN   
1                              NaN                            NaN   
2                              NaN                            NaN   
3                              NaN                            NaN   
4                              NaN                            NaN   

   Total male population raw value  Female population 0-17 raw value  \
0                              NaN                               NaN   
1                              NaN                               NaN   
2                              NaN                               NaN   
3                              NaN                               NaN   
4                              NaN                               NaN   

   Female population 18-44 raw value  Female population 45-64 raw value  \
0                                NaN                                NaN   
1                                NaN                                NaN   
2                                NaN                                NaN   
3                                NaN                                NaN   
4                                NaN                                NaN   

   Female population 65+ raw value  Total female population raw value  \
0                              NaN                                NaN   
1                              NaN                                NaN   
2                              NaN                                NaN   
3                              NaN                                NaN   
4                              NaN                                NaN   

   Population growth raw value  
0                          NaN  
1                          NaN  
2                          NaN  
3                          NaN  
4                          NaN  

[5 rows x 534 columns]

<li>We can see that there are some NAN values and we can replace them with 0's

In [43]:
file2=file2.fillna(0)
file2.head()

State FIPS Code  County FIPS Code  5-digit FIPS Code State Abbreviation  \
0                1                 0               1000                 AL   
1                1                 1               1001                 AL   
2                1                 3               1003                 AL   
3                1                 5               1005                 AL   
4                1                 7               1007                 AL   

             Name  Release Year  County Ranked (Yes=1/No=0)  \
0         Alabama          2019                         0.0   
1  Autauga County          2019                         1.0   
2  Baldwin County          2019                         1.0   
3  Barbour County          2019                         1.0   
4     Bibb County          2019                         1.0   

   Premature death raw value  Premature death numerator  \
0                9917.232898                    80440.0   
1                8824.057123                      815.0   
2                7224.632160                     2827.0   
3                9586.165037                      451.0   
4               11783.543680                      445.0   

   Premature death denominator  ...  Male population 18-44 raw value  \
0                   13636816.0  ...                              0.0   
1                     156132.0  ...                              0.0   
2                     576496.0  ...                              0.0   
3                      72222.0  ...                              0.0   
4                      63653.0  ...                              0.0   

   Male population 45-64 raw value  Male population 65+ raw value  \
0                              0.0                            0.0   
1                              0.0                            0.0   
2                              0.0                            0.0   
3                              0.0                            0.0   
4                              0.0                            0.0   

   Total male population raw value  Female population 0-17 raw value  \
0                              0.0                               0.0   
1                              0.0                               0.0   
2                              0.0                               0.0   
3                              0.0                               0.0   
4                              0.0                               0.0   

   Female population 18-44 raw value  Female population 45-64 raw value  \
0                                0.0                                0.0   
1                                0.0                                0.0   
2                                0.0                                0.0   
3                                0.0                                0.0   
4                                0.0                                0.0   

   Female population 65+ raw value  Total female population raw value  \
0                              0.0                                0.0   
1                              0.0                                0.0   
2                              0.0                                0.0   
3                              0.0                                0.0   
4                              0.0                                0.0   

   Population growth raw value  
0                          0.0  
1                          0.0  
2                          0.0  
3                          0.0  
4                          0.0  

[5 rows x 534 columns]

###### Reading file-3

In [44]:
file3=pd.read_csv('../data/2019-Opioid_Rate.csv')
file3

State             County   FIPS  Opiod_Dispensing_Rate
0       AL     Autauga County   1001                  101.3
1       AL     Baldwin County   1003                   67.6
2       AL     Barbour County   1005                   27.2
3       AL        Bibb County   1007                   21.0
4       AL      Blount County   1009                   23.7
...    ...                ...    ...                    ...
3090    WY  Sweetwater County  56037                   70.0
3091    WY       Teton County  56039                   54.6
3092    WY       Uinta County  56041                   59.5
3093    WY    Washakie County  56043                   46.7
3094    WY      Weston County  56045                    4.7

[3095 rows x 4 columns]

<li>There are 8 columns in file 1 and 534 in file 2 and 4 in file3. Now, we need to examine which columns are necessary for futher analysis

In [45]:
print(file1.columns)

Index(['Notes', 'County', 'County Code', 'Drug/Alcohol Induced Cause',
       'Drug/Alcohol Induced Cause Code', 'Deaths', 'Population',
       'Crude Rate'],
      dtype='object')


<li>From, file1 out of the 8 columns we need County(here county also contains the name of the state, which needs to be split as another column),County code,Drug/Alcohol Induced Cause Code,Deaths,Population.
<li>Now, preparing a data dictionary for the selected columns, 

| Variable Name | Datatype | Description |
| --- | --- | --- |
|County|Object | different county's mapping to the states in the US|
|County code|int| an integer number desctibing certain standard as per US regulations|
|Drug/Alcohol Induced Cause Code| Object | Specifies the cause of death|
|Drug/Alcohol Induced Cause Code| Object | A number assosciated with the Drug Induced cause|
|Deaths|int|an integer number describing the deaths in the country|
|Population| int | an integer number indicating the total population in the country|

In [10]:
print(file2.columns)

Index(['State FIPS Code', 'County FIPS Code', '5-digit FIPS Code',
       'State Abbreviation', 'Name', 'Release Year',
       'County Ranked (Yes=1/No=0)', 'Premature death raw value',
       'Premature death numerator', 'Premature death denominator',
       ...
       'Male population 18-44 raw value', 'Male population 45-64 raw value',
       'Male population 65+ raw value', 'Total male population raw value',
       'Female population 0-17 raw value', 'Female population 18-44 raw value',
       'Female population 45-64 raw value', 'Female population 65+ raw value',
       'Total female population raw value', 'Population growth raw value'],
      dtype='object', length=534)


<li> File2 has 534 columns out of which we only select the<b>10  columns</b> that are needed 5-digit FIPS Code,State Abbreviation,Name,Excessive drinking raw value,Sexually transmitted infections raw value,Mental health providers raw value,Preventable hospital stays raw value,Total male population raw value,Total female population raw value,Population growth raw value.
<li> Now we Create a data dictionary for the columns selected and give the <b>justification.

| Variable Name | Datatype | Description |Justification
| --- | --- | --- |--- |
|5-digit FIPS Code|int| an integer number desctibing certain standard as per US regulations| This number is important during analysis how opioid is spread across different county's.Since each county has a unique FIPS code it becomes easier to group the counties in further analysis|
|State Abbreviation|Object| different states in the US| This field was chosen to understand which county the particular state belongs to|
|Name| Object |  different county's in the US| FIPS number alone might not be enough for analysis, we also need to know to which county those numbers indicate to.|
|Excessive drinking raw value| float |an decimal(float)number describing how the rate of alcohol consumers|This number can give us information if excessive drinking has influence on the opioid mortality rate.|
|Sexually transmitted infections raw value| float |an decimal(float)number describing the rate at which the disease was sexually transmitted|This number can give us information if opioid effects can be sexually tranmitted|
|Mental health providers raw value| float |an decimal(float)number describing mental health assiatance|This is the number of mental health providers in that particular state and county which can be useful during an emergency|
|Preventable hospital stays raw value| float |an decimal(float)number describing the level of disease, whether hospitable or not|This number is important to understand when the patient needs to be hospitalised and when it might not be required.|
|Total male population raw value| float | a decimal number indicating the total male population in the country|This count is important to understand how opioid mortality rates are spread across the male population|
|Total female population raw value|float| a decimal number indicating the total female population in the country|This count is important to understand how opioid mortality rates are spread across the female population|
|Population growth raw value|float|  a decimal number indicating the total population in the country|This count is important to understand how opioid mortality rates are spread across the entire population|

In [11]:
print(file3.columns)

Index(['State', 'County', 'FIPS', 'Opiod_Dispensing_Rate'], dtype='object')


<li> File3 has only 4 columns and we might consider including all the 4 columns
<li>Now, we can create data dictionary for the selected the columns

| Variable Name | Datatype | Description|
| --- | --- | --- |
|State| Object| different states in the US|
|County| Object|  different county's in the US|
|countyFIPS	|int| an integer number describing certain standard as per US regulations|
|Opiod_Dispensing_Rate| float |a decimal(float) number describing the opiod dispensing rate|

### Task M3.1
Merge all the three datasets to create a super datafame.

* First,we select only the necessary columns from file1.
* Next, we normalise the opiod mortality rate per 100000 of the population and create a new column called norm_deaths

In [12]:
file1=file1[['County','County Code','Population','Deaths']]
file1["Norm_Deaths"] = (file1["Deaths"]/file1["Population"])*100000
file1

County  County Code  Population  Deaths  Norm_Deaths
0        Autauga County, AL         1001     1087149      69     6.346876
1        Autauga County, AL         1001     1087149      14     1.287772
2        Baldwin County, AL         1003     3758097     424    11.282306
3        Baldwin County, AL         1003     3758097      71     1.889254
4        Baldwin County, AL         1003     3758097      19     0.505575
...                     ...          ...         ...     ...          ...
5536  Sweetwater County, WY        56037      873221      15     1.717778
5537       Teton County, WY        56039      440125      13     2.953706
5538       Uinta County, WY        56041      426347      51    11.962087
5539       Uinta County, WY        56041      426347      10     2.345507
5540    Washakie County, WY        56043      172114      14     8.134144

[5541 rows x 5 columns]

In [13]:
#renaming the column names in file2 to perform the merge operation

file2.rename(columns = {'State Abbreviation':'State'}, inplace = True)
file2.rename(columns = {'Name':'County'}, inplace = True)
file2.drop(columns=['County Ranked (Yes=1/No=0)'],inplace=True)
file2

State FIPS Code  County FIPS Code  5-digit FIPS Code State  \
0                   1                 0               1000    AL   
1                   1                 1               1001    AL   
2                   1                 3               1003    AL   
3                   1                 5               1005    AL   
4                   1                 7               1007    AL   
...               ...               ...                ...   ...   
3188               56                37              56037    WY   
3189               56                39              56039    WY   
3190               56                41              56041    WY   
3191               56                43              56043    WY   
3192               56                45              56045    WY   

                 County  Release Year  Premature death raw value  \
0               Alabama          2019                9917.232898   
1        Autauga County          2019                8824.057123   
2        Baldwin County          2019                7224.632160   
3        Barbour County          2019                9586.165037   
4           Bibb County          2019               11783.543680   
...                 ...           ...                        ...   
3188  Sweetwater County          2019                7497.439952   
3189       Teton County          2019                3786.128226   
3190       Uinta County          2019                7790.302043   
3191    Washakie County          2019                5504.650970   
3192      Weston County          2019                4991.168273   

      Premature death numerator  Premature death denominator  \
0                       80440.0                   13636816.0   
1                         815.0                     156132.0   
2                        2827.0                     576496.0   
3                         451.0                      72222.0   
4                         445.0                      63653.0   
...                         ...                          ...   
3188                      495.0                     127427.0   
3189                      124.0                      66351.0   
3190                      262.0                      59466.0   
3191                      108.0                      22335.0   
3192                       82.0                      19599.0   

      Premature death CI low  ...  Male population 18-44 raw value  \
0                9815.190945  ...                              0.0   
1                7935.325830  ...                              0.0   
2                6794.128380  ...                              0.0   
3                8200.117541  ...                              0.0   
4               10159.978320  ...                              0.0   
...                      ...  ...                              ...   
3188             6599.947576  ...                              0.0   
3189             2819.782887  ...                              0.0   
3190             6427.289477  ...                              0.0   
3191             3712.604945  ...                              0.0   
3192             3969.624248  ...                              0.0   

      Male population 45-64 raw value  Male population 65+ raw value  \
0                                 0.0                            0.0   
1                                 0.0                            0.0   
2                                 0.0                            0.0   
3                                 0.0                            0.0   
4                                 0.0                            0.0   
...                               ...                            ...   
3188                              0.0                            0.0   
3189                              0.0                            0.0   
3190                              0.0                            0.0   
3191                              0.0     

* We then join the state and the county fields in file2 as in file1 to perform the merge

In [14]:
file2['County'] = file2[['County', 'State']].apply(lambda x: ', '.join(x[x.notnull()]), axis = 1)
file2.head()

State FIPS Code  County FIPS Code  5-digit FIPS Code State  \
0                1                 0               1000    AL   
1                1                 1               1001    AL   
2                1                 3               1003    AL   
3                1                 5               1005    AL   
4                1                 7               1007    AL   

               County  Release Year  Premature death raw value  \
0         Alabama, AL          2019                9917.232898   
1  Autauga County, AL          2019                8824.057123   
2  Baldwin County, AL          2019                7224.632160   
3  Barbour County, AL          2019                9586.165037   
4     Bibb County, AL          2019               11783.543680   

   Premature death numerator  Premature death denominator  \
0                    80440.0                   13636816.0   
1                      815.0                     156132.0   
2                     2827.0                     576496.0   
3                      451.0                      72222.0   
4                      445.0                      63653.0   

   Premature death CI low  ...  Male population 18-44 raw value  \
0             9815.190945  ...                              0.0   
1             7935.325830  ...                              0.0   
2             6794.128380  ...                              0.0   
3             8200.117541  ...                              0.0   
4            10159.978320  ...                              0.0   

   Male population 45-64 raw value  Male population 65+ raw value  \
0                              0.0                            0.0   
1                              0.0                            0.0   
2                              0.0                            0.0   
3                              0.0                            0.0   
4                              0.0                            0.0   

   Total male population raw value  Female population 0-17 raw value  \
0                              0.0                               0.0   
1                              0.0                               0.0   
2                              0.0                               0.0   
3                              0.0                               0.0   
4                              0.0                               0.0   

   Female population 18-44 raw value  Female population 45-64 raw value  \
0                                0.0                                0.0   
1                                0.0                                0.0   
2                                0.0                                0.0   
3                                0.0                                0.0   
4                                0.0                                0.0   

   Female population 65+ raw value  Total female population raw value  \
0                              0.0                                0.0   
1                              0.0                                0.0   
2                              0.0                                0.0   
3                              0.0                                0.0   
4                              0.0                                0.0   

   Population growth raw value  
0                          0.0  
1                          0.0  
2                          0.0  
3                          0.0  
4                          0.0  

[5 rows x 533 columns]

* Once the county field consists of both the county name and state name, we then perofrm the merge operation on file1 and file2 and save it in the final1 variable

In [15]:
final1=file1.merge(file2,on='County',how="left")
final1

County  County Code  Population  Deaths  Norm_Deaths  \
0        Autauga County, AL         1001     1087149      69     6.346876   
1        Autauga County, AL         1001     1087149      14     1.287772   
2        Baldwin County, AL         1003     3758097     424    11.282306   
3        Baldwin County, AL         1003     3758097      71     1.889254   
4        Baldwin County, AL         1003     3758097      19     0.505575   
...                     ...          ...         ...     ...          ...   
5541  Sweetwater County, WY        56037      873221      15     1.717778   
5542       Teton County, WY        56039      440125      13     2.953706   
5543       Uinta County, WY        56041      426347      51    11.962087   
5544       Uinta County, WY        56041      426347      10     2.345507   
5545    Washakie County, WY        56043      172114      14     8.134144   

      State FIPS Code  County FIPS Code  5-digit FIPS Code State  \
0                 1.0               1.0             1001.0    AL   
1                 1.0               1.0             1001.0    AL   
2                 1.0               3.0             1003.0    AL   
3                 1.0               3.0             1003.0    AL   
4                 1.0               3.0             1003.0    AL   
...               ...               ...                ...   ...   
5541             56.0              37.0            56037.0    WY   
5542             56.0              39.0            56039.0    WY   
5543             56.0              41.0            56041.0    WY   
5544             56.0              41.0            56041.0    WY   
5545             56.0              43.0            56043.0    WY   

      Release Year  ...  Male population 18-44 raw value  \
0           2019.0  ...                              0.0   
1           2019.0  ...                              0.0   
2           2019.0  ...                              0.0   
3           2019.0  ...                              0.0   
4           2019.0  ...                              0.0   
...            ...  ...                              ...   
5541        2019.0  ...                              0.0   
5542        2019.0  ...                              0.0   
5543        2019.0  ...                              0.0   
5544        2019.0  ...                              0.0   
5545        2019.0  ...                              0.0   

      Male population 45-64 raw value  Male population 65+ raw value  \
0                                 0.0                            0.0   
1                                 0.0                            0.0   
2                                 0.0                            0.0   
3                                 0.0                            0.0   
4                                 0.0                            0.0   
...                               ...                            ...   
5541                              0.0                            0.0   
5542                              0.0                            0.0   
5543                              0.0                            0.0   
5544                              0.0                            0.0   
5545                              0.0                            0.0   

      Total male population raw value  Female population 0-17 raw value  \
0                                 0.0                               0.0   
1                                 0.0                               0.0   
2                                 0.0                               0.0   
3                                 0.0                               0.0   
4                                 0.0                               0.0   
...                               ...                               ...   
5541                              0.0                               0.0   
5542                              0.0                               0.0   
5543              

* Next, we have to merge this final1 file with the file3 to create the final superdataframe.
* In order to merge file3 with the concatined file1 and file2, we join both county and state fields in file3 and call it as County.

In [16]:
#joining state and county 
file3['County'] = file3[['County', 'State']].apply(lambda x: ', '.join(x[x.notnull()]), axis = 1)
file3=file3[['County','State','FIPS','Opiod_Dispensing_Rate']]
file3

County State   FIPS  Opiod_Dispensing_Rate
0        Autauga County, AL    AL   1001                  101.3
1        Baldwin County, AL    AL   1003                   67.6
2        Barbour County, AL    AL   1005                   27.2
3           Bibb County, AL    AL   1007                   21.0
4         Blount County, AL    AL   1009                   23.7
...                     ...   ...    ...                    ...
3090  Sweetwater County, WY    WY  56037                   70.0
3091       Teton County, WY    WY  56039                   54.6
3092       Uinta County, WY    WY  56041                   59.5
3093    Washakie County, WY    WY  56043                   46.7
3094      Weston County, WY    WY  56045                    4.7

[3095 rows x 4 columns]

* Now file3 and concat both have the county field in common to perform the merge operation, now we can perform the merge operation on the common "County" field.

In [17]:
final=final1.merge(file3,on='County',how="left")
final.drop_duplicates()

County  County Code  Population  Deaths  Norm_Deaths  \
0        Autauga County, AL         1001     1087149      69     6.346876   
1        Autauga County, AL         1001     1087149      14     1.287772   
2        Baldwin County, AL         1003     3758097     424    11.282306   
3        Baldwin County, AL         1003     3758097      71     1.889254   
4        Baldwin County, AL         1003     3758097      19     0.505575   
...                     ...          ...         ...     ...          ...   
5541  Sweetwater County, WY        56037      873221      15     1.717778   
5542       Teton County, WY        56039      440125      13     2.953706   
5543       Uinta County, WY        56041      426347      51    11.962087   
5544       Uinta County, WY        56041      426347      10     2.345507   
5545    Washakie County, WY        56043      172114      14     8.134144   

      State FIPS Code  County FIPS Code  5-digit FIPS Code State_x  \
0                 1.0               1.0             1001.0      AL   
1                 1.0               1.0             1001.0      AL   
2                 1.0               3.0             1003.0      AL   
3                 1.0               3.0             1003.0      AL   
4                 1.0               3.0             1003.0      AL   
...               ...               ...                ...     ...   
5541             56.0              37.0            56037.0      WY   
5542             56.0              39.0            56039.0      WY   
5543             56.0              41.0            56041.0      WY   
5544             56.0              41.0            56041.0      WY   
5545             56.0              43.0            56043.0      WY   

      Release Year  ...  Total male population raw value  \
0           2019.0  ...                              0.0   
1           2019.0  ...                              0.0   
2           2019.0  ...                              0.0   
3           2019.0  ...                              0.0   
4           2019.0  ...                              0.0   
...            ...  ...                              ...   
5541        2019.0  ...                              0.0   
5542        2019.0  ...                              0.0   
5543        2019.0  ...                              0.0   
5544        2019.0  ...                              0.0   
5545        2019.0  ...                              0.0   

      Female population 0-17 raw value  Female population 18-44 raw value  \
0                                  0.0                                0.0   
1                                  0.0                                0.0   
2                                  0.0                                0.0   
3                                  0.0                                0.0   
4                                  0.0                                0.0   
...                                ...                                ...   
5541                               0.0                                0.0   
5542                               0.0                                0.0   
5543                               0.0                                0.0   
5544                               0.0                                0.0   
5545                               0.0                                0.0   

      Female population 45-64 raw value  Female population 65+ raw value  \
0                                   0.0                              0.0   
1                                   0.0                              0.0   
2                                   0.0                              0.0   
3                                   0.0                              0.0   
4                                   0.0                              0.0   
...                                 ...                              ...   
5541                                0.0                              0.0   
5

* Since,there are many counties with repeated values,we groupby the county names and then reset the index.

## Display the super dataframe 

In [18]:
finaldf=final.groupby('County')
finaldf=finaldf.first().reset_index()
finaldf

County  County Code  Population  Deaths  \
0              Abbeville County, SC        45001      535389      38   
1                 Acadia Parish, LA        22001     1279727      98   
2               Accomack County, VA        51001      726189      50   
3                    Ada County, ID        16001     8083452     692   
4                  Adair County, KY        21001      387950      43   
...                             ...          ...         ...     ...   
2525                Yuba County, CA         6115     1468094     112   
2526  Yukon-Koyukuk Census Area, AK         2290      129587      14   
2527                Yuma County, AZ         4027     3964172     381   
2528              Zapata County, TX        48505      284033      16   
2529              Zavala County, TX        48507      247499      16   

      Norm_Deaths  State FIPS Code  County FIPS Code  5-digit FIPS Code  \
0        7.097643             45.0               1.0            45001.0   
1        7.657883             22.0               1.0            22001.0   
2        6.885260             51.0               1.0            51001.0   
3        8.560699             16.0               1.0            16001.0   
4       11.083903             21.0               1.0            21001.0   
...           ...              ...               ...                ...   
2525     7.628939              6.0             115.0             6115.0   
2526    10.803553              2.0             290.0             2290.0   
2527     9.611087              4.0              27.0             4027.0   
2528     5.633148             48.0             505.0            48505.0   
2529     6.464673             48.0             507.0            48507.0   

     State_x  Release Year  ...  Total male population raw value  \
0         SC        2019.0  ...                              0.0   
1         LA        2019.0  ...                              0.0   
2         VA        2019.0  ...                              0.0   
3         ID        2019.0  ...                              0.0   
4         KY        2019.0  ...                              0.0   
...      ...           ...  ...                              ...   
2525      CA        2019.0  ...                              0.0   
2526      AK        2019.0  ...                              0.0   
2527      AZ        2019.0  ...                              0.0   
2528      TX        2019.0  ...                              0.0   
2529      TX        2019.0  ...                              0.0   

      Female population 0-17 raw value  Female population 18-44 raw value  \
0                                  0.0                                0.0   
1                                  0.0                                0.0   
2                                  0.0                                0.0   
3                                  0.0                                0.0   
4                                  0.0                                0.0   
...                                ...                                ...   
2525                               0.0                                0.0   
2526                               0.0                                0.0   
2527                               0.0                                0.0   
2528                               0.0                                0.0   
2529                               0.0                                0.0   

      Female population 45-64 raw value  Female population 65+ raw value  \
0                                   0.0                              0.0   
1                                   0.0                              0.0   
2                                   0.0                              0.0   
3                                   0.0                              0.0   
4                                   0.0                              0.0   
...                                 ...                          

* The finaldf represents the superdataframe, which of shape <b>(2530,541)
    
    
## Export it to a csv format.

In [19]:
finaldf.to_csv('../data/finaldf-raveena.csv')

* I have exported the generate final dataframe into csv and have stored it in the ../data location

## Task M3.2 
### Identify counties and states (top 10) with the highest opioid mortality rates.

In [20]:
#displaying the top-10 states with highest mortality rates
top=finaldf.groupby('State_x').first().reset_index()
top10=top.nlargest(n=10, columns=['Norm_Deaths']) 
top10

State_x                    County  County Code  Population  Deaths  \
35      OH          Adams County, OH        39001      590060     140   
32      NM     Bernalillo County, NM        35001    13368036    3078   
42      TN       Anderson County, TN        47001     1555370     325   
31      NJ       Atlantic County, NJ        34001     5630056    1134   
7       DC  District of Columbia, DC        11001    12923833    2141   
30      NH        Belknap County, NH        33001     1254278     199   
19      MA     Barnstable County, MA        25001     4579885     674   
33      NV           Carson City, NV        32510     1150738     168   
8       DE           Kent County, DE        10001     3274264     453   
5       CO          Adams County, CO         8001     9146345    1170   

    Norm_Deaths  State FIPS Code  County FIPS Code  5-digit FIPS Code  \
35    23.726401             39.0               1.0            39001.0   
32    23.025073             35.0               1.0            35001.0   
42    20.895350             47.0               1.0            47001.0   
31    20.141896             34.0               1.0            34001.0   
7     16.566293             11.0               0.0            11000.0   
30    15.865701             33.0               1.0            33001.0   
19    14.716527             25.0               1.0            25001.0   
33    14.599327             32.0             510.0            32510.0   
8     13.835170             10.0               1.0            10001.0   
5     12.791995              8.0               1.0             8001.0   

    Release Year  ...  Total male population raw value  \
35        2019.0  ...                              0.0   
32        2019.0  ...                              0.0   
42        2019.0  ...                              0.0   
31        2019.0  ...                              0.0   
7         2019.0  ...                              0.0   
30        2019.0  ...                              0.0   
19        2019.0  ...                              0.0   
33        2019.0  ...                              0.0   
8         2019.0  ...                              0.0   
5         2019.0  ...                              0.0   

    Female population 0-17 raw value  Female population 18-44 raw value  \
35                               0.0                                0.0   
32                               0.0                                0.0   
42                               0.0                                0.0   
31                               0.0                                0.0   
7                                0.0                                0.0   
30                               0.0                                0.0   
19                               0.0                                0.0   
33                               0.0                                0.0   
8                                0.0                                0.0   
5                                0.0                                0.0   

    Female population 45-64 raw value  Female population 65+ raw value  \
35                                0.0                              0.0   
32                                0.0                              0.0   
42                                0.0                              0.0   
31                                0.0                              0.0   
7                                 0.0                              0.0   
30                                0.0                              0.0   
19                                0.0                              0.0   
33                                0.0                              0.0   
8                                 0.0                              0.0   
5                                 0.0                              0.0   

    Total female population raw value  Population growth raw value  State_y  \
35                                0.

### The above dataframe displays the top-15 values with the highest mortality rates all over the US.
* From the above dataframe we can see that the <b>Adams County, OH has the highest moratlity rate. 

### Analysing how the mortality rate varies across the states in the top 10 counties

* Making a list of top 10 states with the highest moratlity rates and calculating the mean and median of the death count


In [21]:
top10['State_x'].unique()

array(['OH', 'NM', 'TN', 'NJ', 'DC', 'NH', 'MA', 'NV', 'DE', 'CO'],
      dtype=object)

## Use mean and median for counties within states to compare (for the state level).

In [53]:
import plotly.express as px
statelist=['OH', 'NM', 'TN', 'NJ', 'DC', 'NH', 'MA', 'NV', 'DE', 'CO']
for i,x in enumerate(statelist):
    state=finaldf[finaldf.State_x==(statelist[i])]
    print("\033[1m"+statelist[i],"\033[0m")
    print("The mean of opioid mortality in\t" +statelist[i],":",state['Deaths'].mean())
    print("The median of opioid mortality in\t" +statelist[i],":",state['Deaths'].median())
    print("\n")
    county=state[state['Norm_Deaths']==state['Norm_Deaths'].max()]
    print("The county with maximum moratality rate in state:"+statelist[i] ,"is-",county['County'])
    print("With a death(moratlity) count of:",county['Deaths'])
    print("---------------------------------------------------------------------------------")

OH 
The mean of opioid mortality in	OH : 447.35227272727275
The median of opioid mortality in	OH : 163.0


The county with maximum moratality rate in state:OH is- 1597    Montgomery County, OH
Name: County, dtype: object
With a death(moratlity) count of: 1597    3523
Name: Deaths, dtype: int64
---------------------------------------------------------------------------------
NM 
The mean of opioid mortality in	NM : 253.73333333333332
The median of opioid mortality in	NM : 101.0


The county with maximum moratality rate in state:NM is- 1957    Rio Arriba County, NM
Name: County, dtype: object
With a death(moratlity) count of: 1957    477
Name: Deaths, dtype: int64
---------------------------------------------------------------------------------
TN 
The mean of opioid mortality in	TN : 198.0
The median of opioid mortality in	TN : 87.0


The county with maximum moratality rate in state:TN is- 1099    Jackson County, TN
Name: County, dtype: object
With a death(moratlity) count of: 1099    6

* The mean and median for the top 10 affected state are as displayed above.If the mean and median are around same value the we can that it follows a uniform distribution.

### Plotting the Deaths across each of these 10 states to see how the data is spread and displaying the top-10 counties with the highest mortality rate in each of these states

In [57]:
for i,x in enumerate(statelist):
    state=finaldf[finaldf.State_x==(statelist[i])]
    fig = px.histogram(state, x="County", y="Norm_Deaths", color="Population",hover_name="Population", title ='Opiod Mortality rate in\t'+statelist[i])
    fig.show()
    print("The Number of counties in-"+statelist[i],":",len(state['County']))
    tencounty=state.nlargest(n=10, columns=['Norm_Deaths'])
    #print(tencounty['County'])
    print("Top-10 counties with highest mortatlity rate in state-"+statelist[i],"are:\n",tencounty['County'])

The Number of counties in-OH : 88
Top-10 counties with highest mortatlity rate in state-OH are:
 1597    Montgomery County, OH
2051        Scioto County, OH
244          Brown County, OH
276         Butler County, OH
1987          Ross County, OH
12           Adams County, OH
2270      Trumbull County, OH
469       Clermont County, OH
479        Clinton County, OH
440          Clark County, OH
Name: County, dtype: object


The Number of counties in-NM : 30
Top-10 counties with highest mortatlity rate in state-NM are:
 1957    Rio Arriba County, NM
168     Bernalillo County, NM
2026    San Miguel County, NM
916      Guadalupe County, NM
2101        Sierra County, NM
2259      Torrance County, NM
1335       Lincoln County, NM
2311      Valencia County, NM
2038      Santa Fe County, NM
687           Eddy County, NM
Name: County, dtype: object


The Number of counties in-TN : 94
Top-10 counties with highest mortatlity rate in state-TN are:
 1099      Jackson County, TN
459          Clay County, TN
1964        Roane County, TN
160        Benton County, TN
393      Cheatham County, TN
314      Campbell County, TN
1216         Knox County, TN
610        DeKalb County, TN
956        Hardin County, TN
429     Claiborne County, TN
Name: County, dtype: object


The Number of counties in-NJ : 21
Top-10 counties with highest mortatlity rate in state-NJ are:
 309         Camden County, NJ
851     Gloucester County, NJ
86        Atlantic County, NJ
2005         Salem County, NJ
322       Cape May County, NJ
568     Cumberland County, NJ
1697         Ocean County, NJ
719          Essex County, NJ
269     Burlington County, NJ
1562      Monmouth County, NJ
Name: County, dtype: object


The Number of counties in-DC : 1
Top-10 counties with highest mortatlity rate in state-DC are:
 643    District of Columbia, DC
Name: County, dtype: object


The Number of counties in-NH : 10
Top-10 counties with highest mortatlity rate in state-NH are:
 1016    Hillsborough County, NH
2179       Strafford County, NH
144          Belknap County, NH
525             Coos County, NH
338          Carroll County, NH
1978      Rockingham County, NH
1524       Merrimack County, NH
406         Cheshire County, NH
2186        Sullivan County, NH
864          Grafton County, NH
Name: County, dtype: object


The Number of counties in-MA : 14
Top-10 counties with highest mortatlity rate in state-MA are:
 231        Bristol County, MA
718          Essex County, MA
112     Barnstable County, MA
1849      Plymouth County, MA
2181       Suffolk County, MA
936        Hampden County, MA
167      Berkshire County, MA
2495     Worcester County, MA
1682       Norfolk County, MA
671          Dukes County, MA
Name: County, dtype: object


The Number of counties in-NV : 13
Top-10 counties with highest mortatlity rate in state-NV are:
 1694           Nye County, NV
2445    White Pine County, NV
439          Clark County, NV
2400        Washoe County, NV
425      Churchill County, NV
342           Carson City, NV
1546       Mineral County, NV
1388          Lyon County, NV
662        Douglas County, NV
1257        Lander County, NV
Name: County, dtype: object


The Number of counties in-DE : 3
Top-10 counties with highest mortatlity rate in state-DE are:
 1653    New Castle County, DE
2202        Sussex County, DE
1186          Kent County, DE
Name: County, dtype: object


The Number of counties in-CO : 46
Top-10 counties with highest mortatlity rate in state-CO are:
 1264    Las Animas County, CO
1045      Huerfano County, CO
517        Conejos County, CO
632         Denver County, CO
1959    Rio Grande County, CO
532       Costilla County, CO
1894        Pueblo County, CO
7            Adams County, CO
22         Alamosa County, CO
696        El Paso County, CO
Name: County, dtype: object


* I have also displayed the number of counties in each of these states so we can clearly see how the mortality rate is spread across each of these states among their counties.
* Top-10 counties in each of these states are also displayed at the bottom of the graph.
    * Since, DC has no counties, the graph shows a single county.
    * And, there are only 3 counties in DE-delaware which is as shown in the above graph

## Describe your intution on why the rates are high in these states and counties.

* From the above plotted graphs we can see that the mortality rate is propotional to population. <b>In the counties where the population is high,the mortality rate is also high.</b>
* Also, I believe that the opiod_dispensing_rate can high a direct influence on the mortality rate.
* And there can be a scope for comparission between the age groups of the population, are the rates high among the teens or adults?